src/stream/stream.rs

```diff
@@ -273,3 +330,50 @@ where
         }
     }
 }
+
+#[derive(Debug)]
+pub struct AnyFuture<'a, S, F, T>
+where
+    F: FnMut(T) -> bool,
+{
+    stream: &'a mut S,
+    f: F,
+    result: bool,
+    __item: PhantomData<T>,
+}
+
+impl<'a, S, F, T> AnyFuture<'a, S, F, T>
+where
+    F: FnMut(T) -> bool,
+{
+    pin_utils::unsafe_pinned!(stream: &'a mut S);
+    pin_utils::unsafe_unpinned!(result: bool);
+    pin_utils::unsafe_unpinned!(f: F);
+}
+
+impl<S, F> Future for AnyFuture<'_, S, F, S::Item>
+where
+    S: futures::Stream + Unpin + Sized,
+    F: FnMut(S::Item) -> bool,
+{
+    type Output = bool;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        use futures::Stream;
+        let next = futures::ready!(self.as_mut().stream().poll_next(cx));
+        match next {
+            Some(v) => {
+                let result = (self.as_mut().f())(v);
+                *self.as_mut().result() = result;
```

Comment:
why we need a `result` field in `AnyFuture` and update every time here, instead:
```rust
        match next {
            Some(v) => {
                let result = (self.as_mut().f())(v);
                if result {
                    Poll::Ready(true)
                } else {
                    // don't forget to wake this task again to pull the next item from stream
                    cx.waker().wake_by_ref();
                    Poll::Pending
                }
            }
            None => Poll::Ready(false),
        }
```

Reply:
We should save the last result somewhere since the stream would yield `None` in two cases if it's already an empty stream or we get the end of the stream.

in the first case, the default value of `result` is `fasle` .
in the second case, we should return the last `result` when we get `None`.

in your example, the result of `stream::any` method would always return `false` whatever the stream is empty or have elements on it.

Reply:
> in the second case, we should return the last result when we get None.

if the last `result` is `true`, does not it return `Poll::Ready(true)` directly in line 369? so if it reaches the `None` case, `self.result` should be `false`.

Reply:
If it returned with `Poll::Ready` then we are done, the Future is resolved and there is no calls to the `poll` method again.
So no, it can't reach the None case here, since the `any` method is short-circuit it will stop once it got `true`.